p2p/peer.go

```diff
@@ -239,27 +247,41 @@ const (
 // - Disconnected: peer disconnects, unmarking as connected and broadcasts a
 //   PeerStatusDown peer update.
 //
-// If we need to evict a peer, typically because we have connected to additional
-// higher-scored peers and need to shed lower-scored ones, the flow is as follows:
+// If we are connected to too many peers (more than MaxConnections), typically
+// because we have upgraded to higher-scored peers and need to shed lower-scored
+// ones, the flow is as follows:
 // - EvictNext: returns a peer ID to evict, marking peer as evicting.
 // - Disconnected: peer was disconnected, unmarking as connected and evicting,
 //   and broadcasts a PeerStatusDown peer update.
 //
-// We track dialing and connected states independently. This allows us to accept
-// an inbound connection from a peer while the router is also dialing an
-// outbound connection to that same peer, which will cause the dialer to
-// eventually error (when attempting to mark the peer as connected). This also
-// avoids race conditions where multiple goroutines may end up dialing a peer if
-// an incoming connection was briefly accepted and disconnected while we were
-// also dialing.
+// If all connection slots are full (at MaxConnections), we can use up to
+// MaxConnectionsUpgrade additional connections to probe any higher-scored
+// unconnected peers, and if we reach them (or they reach us) we allow the
+// connection and evict lower-scored peers. We mark the lower-scored peer as
+// upgrading[from]=to to make sure no other higher-scored peers can claim the
+// same one for an upgrade. The flow is as follows:
+// - Accepted: if upgrade is possible, mark upgrading[from]=to and connected.
+// - DialNext: if upgrade is possible, mark upgrading[from]=to and dialing.
+// - DialFailed: unmark upgrading[from]=to and dialing.
+// - Dialed: unmark dialing, mark as connected.
+// - EvictNext: unmark upgrading[from]=to, then if over MaxConnections
+//   either the upgraded peer or an even lower-scored one (if found)
+//   is marked as evicting and returned.
```

Comment:
Why is it possible for the upgraded peer to be marked as evicted? Shouldn't it just be the lower-scored peer?

Reply:
Yeah, "upgraded peer" is an ambiguous term -- I've tried to be careful to only use it to refer to the lower-scored peer that has been upgraded, not the higher-scored one that was upgraded to. Will try to reword throughout.

Reply:
Ahhh I see. Yeah I think if we can, we should strive to make it less ambiguous.